src/find/matchers/logical_matchers.rs

```diff
@@ -0,0 +1,477 @@
+//! This modules contains the matchers used for combining other matchers and
+//! performing boolean logic on them (and a couple of trivial always-true and
+//! always-false matchers). The design is strongly tied to the precedence rules
+//! when parsing command-line options (e.g. "-foo -o -bar -baz" is equivalent
+//! to "-foo -o ( -bar -baz )", not "( -foo -o -bar ) -baz").
+use std::error::Error;
+use std::iter::Iterator;
+use walkdir::DirEntry;
+
+use find::matchers::{Matcher, MatcherIO};
+
+/// This matcher contains a collection of other matchers. A file only matches
+/// if it matches ALL the contained sub-matchers. For sub-matchers that have
+/// side effects, the side effects occur in the same order as the sub-matchers
+/// were pushed into the collection.
+pub struct AndMatcher {
+    submatchers: Vec<Box<Matcher>>,
```

Comment:
Vec<Matcher> should work ok here, not sure why you're adding an additional `Box` indirection.